File: backend/app/schemas/boss.py
from datetime import date
from pydantic import BaseModel, Field


class BossCurrentOut(BaseModel):
    available: bool
    week_start: date
    snapshot_token: str | None = None
    challenge: dict | None = None
    best_score: int | None = None
    best_medal: str | None = None


class BossCompleteIn(BaseModel):
    snapshot_token: str = Field(min_length=20)
    idempotency_key: str = Field(min_length=8, max_length=128)
    vocabulary_correct: int = Field(ge=0, le=10)
    reading_correct: int = Field(ge=0, le=10)
    listening_correct: int = Field(ge=0, le=10)
    speaking_score: int | None = Field(default=None, ge=0, le=100)
    duration_seconds: int = Field(default=0, ge=0, le=1800)


class BossCompleteOut(BaseModel):
    score: int
    medal: str | None
    best_score: int
    best_medal: str | None
    xp_awarded: int
    unlocks: list[str]
    replay_available: bool = True

File: frontend/src/api/boss.ts
import client from './client'
export type BossCurrent = { available: boolean; week_start: string; snapshot_token?: string; challenge?: { duration_minutes: number; vocabulary_card_ids: string[] }; best_score?: number | null; best_medal?: string | null }
export type BossResult = { score: number; medal: string | null; best_score: number; best_medal: string | null; xp_awarded: number; unlocks: string[] }
export const getCurrentBoss = () => client.get<BossCurrent>('/boss/current').then(result => result.data)
export const completeBoss = (body: { snapshot_token: string; idempotency_key: string; vocabulary_correct: number; reading_correct: number; listening_correct: number; speaking_score?: number | null; duration_seconds: number }) => client.post<BossResult>('/boss/complete', body).then(result => result.data)
